tezos/new_context/src/working_tree/serializer.rs

```diff
@@ -0,0 +1,591 @@
+}
+
+impl From<FromUtf8Error> for DeserializationError {
+    fn from(_: FromUtf8Error) -> Self {
+        Self::FromUtf8Error
+    }
+}
+
+impl From<NodeIdError> for DeserializationError {
+    fn from(_: NodeIdError) -> Self {
+        Self::NodeIdError
+    }
+}
+
+impl From<StorageIdError> for DeserializationError {
+    fn from(error: StorageIdError) -> Self {
+        Self::StorageIdError { error }
+    }
+}
+
+pub fn deserialize(data: &[u8], storage: &mut Storage) -> Result<Entry, DeserializationError> {
+    let mut pos = 1;
+
+    use DeserializationError as Error;
+    use DeserializationError::*;
+
+    match data.get(0).copied().ok_or(UnexpectedEOF)? {
+        ID_TREE => {
+            let data_length = data.len();
+
+            let tree_id =
+                storage.with_new_tree::<_, Result<_, Error>>(|storage, new_tree| {
+                    while pos < data_length {
+                        let descriptor = data.get(pos..pos + 1).ok_or(UnexpectedEOF)?;
+                        let descriptor = KeyNodeDescriptor::from_bytes([descriptor[0]; 1]);
+
+                        pos += 1;
+
+                        let key_id = match descriptor.key_inline_length() as usize {
+                            len if len > 0 => {
+                                // The key is in the next `len` bytes
+                                let key_bytes = data.get(pos..pos + len).ok_or(UnexpectedEOF)?;
+                                let key_str = std::str::from_utf8(key_bytes)?;
+                                pos += len;
+                                storage.get_string_id(key_str)
+                            }
+                            _ => {
+                                // The key length is in 2 bytes, followed by the key itself
+                                let key_length = data.get(pos..pos + 2).ok_or(UnexpectedEOF)?;
+                                let key_length = u16::from_ne_bytes(key_length.try_into()?);
+                                let key_length = key_length as usize;
+
+                                let key_bytes = data
+                                    .get(pos + 2..pos + 2 + key_length)
+                                    .ok_or(UnexpectedEOF)?;
+                                let key_str = std::str::from_utf8(key_bytes)?;
+                                pos += 2 + key_length;
+                                storage.get_string_id(key_str)
+                            }
+                        };
+
+                        let kind = descriptor.kind();
+                        let blob_inline_length = descriptor.blob_inline_length() as usize;
+
+                        let node = if blob_inline_length > 0 {
+                            // The blob is inlined
+
+                            let blob = data
+                                .get(pos..pos + blob_inline_length)
+                                .ok_or(UnexpectedEOF)?;
+                            let blob_id = storage.add_blob_by_ref(blob)?;
```

Comment:
Ok, I see this is why `&mut Storage` is needed here (and elsewhere). But why is deserialization adding new values?

Reply:
deserialization requires `&mut Storage` because it extracts values from the serialized bytes, such as the strings, `Node`, blobs, etc..
The working tree is able to work with values from `Storage` only, not from serialized bytes directly.
I will add some comments